### PR TITLE
Sea Turtles Tyrah G. -Swap Meet

### PR DESCRIPTION
**Pull request by ursaturnine**

### CODE REVIEW

File: tests/unit_tests/test_wave_01.py [hunk 1]
```diff
@@ -49,7 +49,8 @@ def test_removing_not_found_is_false():
 
     result = vendor.remove(item)
 
-    raise Exception("Complete this test according to comments below.")
+    # raise Exception("Complete this test according to comments below.")
     # *********************************************************************
     # ****** Complete Assert Portion of this test **********
     # *********************************************************************
+    assert result == False
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_01.py:56]: 👍 

I like that you explicitly checked for `False` rather than doing amore falsy check since the function says it should literally return `False`.

One thing that would be good still to `assert` is that each of the values in inventory are still there in the list after the `remove` call (since it _usually_ modifies the inventory, but _shouldn't_ in this case).

File: tests/unit_tests/test_wave_02.py [hunk 2]
```diff
@@ -34,7 +34,8 @@ def test_get_no_matching_items_by_category():
 
     items = vendor.get_by_category("electronics")
 
-    raise Exception("Complete this test according to comments below.")
+    # raise Exception("Complete this test according to comments below.")
     # *********************************************************************
     # ****** Complete Assert Portion of this test **********
     # *********************************************************************
+    assert items == []
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_02.py:41]: 👍 

This always looks strange to me since I always want to write empty list checks as `not items`, but here in the tests we want to be specific like this. In application Python code, we should use the more permissive checks (`not items`), but in a test, we want to ensure that the thing returned really is a list and that it is empty.

File: tests/unit_tests/test_wave_06.py [hunk 3]
```diff
@@ -76,16 +76,19 @@ def test_swap_best_by_category():
...
     # *********************************************************************
     # ****** Complete Assert Portion of this test **********
     # *********************************************************************
     # Assertions should check:
     # - That the results is truthy
     # - That tai and jesse's inventories are the correct length
     # - That all the correct items are in tai and jesse's inventories, including the items which were swapped from one vendor to the other
+    assert result == True
+    assert len(tai.inventory) == 3 and len(jesse.inventory) == 3
+    assert  tai.inventory==[item_a, item_b, item_f] and jesse.inventory == [item_d, item_e, item_c]
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_06.py:89]: 👍 

The specification doesn't stipulate where the swapped items will end up in the output. Consider using `in` checks to avoid implying an ordering.

File: tests/unit_tests/test_wave_06.py [hunk 4]
```diff
@@ -109,16 +112,23 @@ def test_swap_best_by_category_reordered():
...
     # - That tai and jesse's inventories are the correct length
     # - That all the correct items are in tai and jesse's inventories, and that the items that were swapped are not there
-
-@pytest.mark.skip
+    assert result == True
+    assert len(tai.inventory) == 3 and len(jesse.inventory) ==3
+    for item in [item_a, item_b, item_f]:
+        assert item in tai.inventory
+    for item in [item_c, item_e, item_d]:
+        assert item in jesse.inventory
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_06.py:128]: 👍 

Ooh! Yes. I love the `in` checks here, and nice use of the loops to minimize repetition!

File: tests/unit_tests/test_wave_06.py [hunk 5]
```diff
@@ -194,16 +204,23 @@ def test_swap_best_by_category_no_match_is_false():
...
     # Assertions should check:
     # - That result is falsy
     # - That tai and jesse's inventories are the correct length
     # - That all the correct items are in tai and jesse's inventories
+    assert result == False
+    assert len(tai.inventory) == 3 and len(jesse.inventory) == 3
+    for item in [item_a, item_b, item_c]:
+        assert item in tai.inventory
+    for item in [item_f, item_e, item_d]:
+        assert item in jesse.inventory
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_06.py:220]: 👍

File: tests/unit_tests/test_wave_06.py [hunk 6]
```diff
@@ -227,11 +244,18 @@ def test_swap_best_by_category_no_other_match_is_false():
...
     # Assertions should check:
     # - That result is falsy
     # - That tai and jesse's inventories are the correct length
     # - That all the correct items are in tai and jesse's inventories
+    assert result == False
+    assert len(tai.inventory) ==3 and len(jesse.inventory) == 3
+    for item in [item_a, item_b, item_c]:
+        assert item in tai.inventory
+    for item in [item_f, item_e, item_d]:
+        assert item in jesse.inventory
```
Comment (anselrognlie) [inline tests/unit_tests/test_wave_06.py:260]: 👍

File: tests/integration_tests/test_wave_01_02_03.py [hunk 7]
```diff
@@ -2,8 +2,8 @@
 from swap_meet.vendor import Vendor
 from swap_meet.item import Item
 
-@pytest.mark.skip
-@pytest.mark.integration_test
+# @pytest.mark.skip
+# @pytest.mark.integration_test
```
Comment (anselrognlie) [inline tests/integration_tests/test_wave_01_02_03.py:6]: We don't need to remove the integration test marker. Removing it didn't hurt anything in this case, but the tag can be used to help tell pytest what kinds of tests to run at certain times.

File: tests/integration_tests/test_wave_04_05_06.py [hunk 8]
```diff
@@ -4,8 +4,8 @@
 from swap_meet.decor import Decor
 from swap_meet.electronics import Electronics
 
-@pytest.mark.skip
-@pytest.mark.integration_test
+# @pytest.mark.skip
+# @pytest.mark.integration_test
```
Comment (anselrognlie) [inline tests/integration_tests/test_wave_04_05_06.py:8]: Here as well, no need to remove this decorator. Just the skip.

File: swap_meet/item.py [hunk 9]
```diff
@@ -1,2 +1,21 @@
+
+
 class Item:
-    pass
\ No newline at end of file
+    def __init__(self, condition=0, category=""):
```
Comment (anselrognlie) [inline swap_meet/item.py:4]: 👍

File: swap_meet/item.py [hunk 10]
```diff
@@ -1,2 +1,21 @@
...
+        self.category = category
+        self.condition = condition
+
+    def __str__(self):
```
Comment (anselrognlie) [inline swap_meet/item.py:8]: 👍

File: swap_meet/item.py [hunk 11]
```diff
@@ -1,2 +1,21 @@
...
+        return "Hello World!"
+
+    def condition_description(self):
```
Comment (anselrognlie) [inline swap_meet/item.py:11]: 👍 

What would be returned if the condition were say -1, or how about 10.

Also, notice how repetitive all the conditions are. If we wanted to add additional conditions, we would need to write more code to handle them. What if we stored the conditions in a collection and could somehow calculate where in the collection the appropriate description for each condition could be found? If we do it carefully, adding additional conditions could be as simple as adding a new string to the collection, and there might be less code overall that would be easier to test (are ALL of these messages currently being tested?).

File: swap_meet/clothing.py [hunk 12]
```diff
@@ -1,2 +1,8 @@
-class Clothing:
-    pass
\ No newline at end of file
+from .item import Item
+
+class Clothing(Item):
+    def __init__(self, condition=0):
+        super().__init__(condition, category="Clothing")
```
Comment (anselrognlie) [inline swap_meet/clothing.py:5]: Nice job  removing the category from the constructor, since we don't want to encourager the caller to try to change it. Instead, we pass the literal value that we want to set up to the parent's constructor. We could also choose to use named parameters as
```python
        super().__init__(category="Clothing", condition=condition)
```

File: swap_meet/clothing.py [hunk 13]
```diff
@@ -1,2 +1,8 @@
...
+
+    def __str__(self):
```
Comment (anselrognlie) [inline swap_meet/clothing.py:7]: 👍

File: swap_meet/vendor.py [hunk 14]
```diff
@@ -1,2 +1,63 @@
+
 class Vendor:
-    pass
\ No newline at end of file
+    def __init__(self, inventory=None):
+        if inventory is None:
```
Comment (anselrognlie) [inline swap_meet/vendor.py:4]: Nice handling of the default condition. I like the explicit `None` check, rather than a more generic falsy check, since we are using `None` specifically as a "sentinel" value here to mark that no value was given.

File: swap_meet/vendor.py [hunk 15]
```diff
@@ -1,2 +1,63 @@
...
+            self.inventory = []
+        else:
+            self.inventory = inventory
+
+    def add(self, item):
```
Comment (anselrognlie) [inline swap_meet/vendor.py:9]: 👍

File: swap_meet/vendor.py [hunk 16]
```diff
@@ -1,2 +1,63 @@
...
+        self.inventory.append(item)
+        return item
+    
+    def remove(self, item):
+        if item in self.inventory:
+            self.inventory.remove(item)
+            return item
+        else:
+            return False
```
Comment (anselrognlie) [inline swap_meet/vendor.py:18]: Nice check to make sure the value is there before we try to remove it. Note that we could instead use try/except to handle the case where `remove` raises an error when the value isn't found.

As written, I'd suggest doing the check for the value _not_ being here first, and then returning as a guard clause, as this allows the main focus of the code to be less indented.

```python
    def remove(self, item):
        if item not in self.inventory:
            return False

        self.inventory.remove(item)
        return item
```

File: swap_meet/vendor.py [hunk 17]
```diff
@@ -1,2 +1,63 @@
...
+
+    def get_by_category(self, category):
+        items_matching_categories = []
+        for item in self.inventory:
+            if item.category == category:
+                items_matching_categories.append(item)
```
Comment (anselrognlie) [inline swap_meet/vendor.py:24]: Notice this pattern of 
```python
result_list = []
for element in source_list:
    if some_condition(element):
        result_list.append(element)
```

can be rewritten as a list comprehension as
```python
result_list = [element for element in source_list if some_condition(element)]
```

Which here would look like
```python
        items_matching_categories = [item for item in self.inventory if item.category == category]
```

At first, this may seem more difficult to read, but comprehensions are very common python style, so try getting used to working with them!

File: swap_meet/vendor.py [hunk 18]
```diff
@@ -1,2 +1,63 @@
...
+
+    def swap_items(self, friend, my_item, their_item):
+        if my_item in self.inventory and their_item in friend.inventory:
+            self.inventory.remove(my_item)
+            self.inventory.append(their_item)
+            friend.inventory.append(my_item)
+            friend.inventory.remove(their_item)
+            return True
+        else:
+            return False
```
Comment (anselrognlie) [inline swap_meet/vendor.py:35]: Here again, I would tend to invert the condition to turn this into a guard clause scenario. Compound conditions can be a little tricky to invert, but we could start with:

```python
        # if it's not the case that both vendors have their needed items
        if not (my_item in self.inventory and their_item in friend.inventory):
```

or a little more cleanly as
```python
        # if either vendor doesn't have their needed item
        if my_item not in self.inventory or their_item not in friend.inventory:
```

Notice how when moving the `not` into each partial condition, the `and` became an `or`. A similar change happens in the way we describe the situations in English.

Then we can write the whole body as 
```python
        if my_item not in self.inventory or their_item not in friend.inventory:
            return False
        
        friend.inventory.append(my_item)
        self.inventory.remove(my_item)
        self.inventory.append(their_item)
        friend.inventory.remove(their_item)
        return True 
```

File: swap_meet/vendor.py [hunk 19]
```diff
@@ -1,2 +1,63 @@
...
+
+    def swap_first_item(self, friend):
+        if self.inventory and friend.inventory:
+            if self.swap_items(friend, self.inventory[0], friend.inventory[0]):
+                return True
+        else:
+            return False
```
Comment (anselrognlie) [inline swap_meet/vendor.py:42]: We definitely need to make sure the inventories aren't empty, otherwise the `[0]` lookup could crash. Again, we can invert the check into a guard clause. We can also observe that the return value of `swap_first_item` is expected to behave the same as `swap_items`, so we can return the result from `swap_items` directly.

```python
    def swap_first_item(self, friend):
        if not self.inventory or not friend.inventory:
            return False

        return self.swap_items(friend, self.inventory[0], friend.inventory[0]):
```

Notice that due to the remove operations in `swap_items`, it is a linear operation, regardless of where in the list each item is. We could achieve constant time here either by swapping the items directly, or by doing an `index` then swap approach in `swap_items`.

File: swap_meet/vendor.py [hunk 20]
```diff
@@ -1,2 +1,63 @@
...
+        
+    def get_best_by_category(self, category):
+        items_in_category = [item for item in self.inventory if item.category == category]
```
Comment (anselrognlie) [inline swap_meet/vendor.py:45]: Nice list comprehension here. Also, don't forget that you have already written a function that returns a list of items that match a category. 🤔

File: swap_meet/vendor.py [hunk 21]
```diff
@@ -1,2 +1,63 @@
...
+        if len(items_in_category) == 0:
+            return None
+        
+        best_condition_item = items_in_category[0]
+        for item in items_in_category:
+            if item.condition > best_condition_item.condition:
+                best_condition_item = item
```
Comment (anselrognlie) [inline swap_meet/vendor.py:52]: Nice logic for finding the item with the best condition!

File: swap_meet/vendor.py [hunk 22]
```diff
@@ -1,2 +1,63 @@
...
+
+        return best_condition_item
+
+    def swap_best_by_category(self, other, my_priority, their_priority):
+        if self.get_best_by_category(their_priority) and other.get_best_by_category(my_priority):
+            self.swap_items(other, self.get_best_by_category(their_priority), other.get_best_by_category(my_priority))
+            return True
+        else:
+            return False
```
Comment (anselrognlie) [inline swap_meet/vendor.py:61]: Nice job reusing `get_best_by_category` and `swap_items`.

Here, I'd suggest storing the results of the first calls to `get_best_by_category` for use later on. Calling them twice isn't free, plus we might make a mistake in the way we call each the second time. So something like this (note also the use of a guard clause).

```python
    def swap_best_by_category(self, other, my_priority, their_priority):
        my_item = self.get_best_by_category(their_priority)
        their_item = other.get_best_by_category(my_priority)

        if not my_item or not their_item:
            return False

        return self.swap_items(other, my_item, their_item)
```

We could even observe that the behavior of `swap_best_by_category` is the same as `swap_items` (in how it returns `True` and `False`). Would that allow us to simplify the code even further? What would `swap_items` do if either input happened to be `None`?

**Review by anselrognlie — COMMENTED** [review]
Nice job!

All required waves are complete and all tests are passing (including the tests which need completion, and I really appreciated how you wrote the `in` assertions for the swap tests)! Your code is well organized, and you are using good, descriptive variable names. You had great code reuse, and appear well on your way with OOP.

The main things I would recommend focusing are some of the smaller details. Watch for places where your code could be a little more pythonic (guard clauses to reduce nesting, list comprehensions). Also, try looking for patterns of repeated code, and even though a helper function might not be a good fit, think about whether putting the repeated data into a collection of some kind might help us write code that's more DRY (and data-driven!).

Overall, well done!
